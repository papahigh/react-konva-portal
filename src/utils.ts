import { ReactNode, useReducer } from 'react';
import { ManagerCommand, MountCommand, PortalMeta, UnmountCommand, UpdateCommand } from './types';

export const Z_INDEX = 0;
export const PORTAL_LAYER_ID = 'react-konva-portal';
export const ERROR_MESSAGE =
  'You should wrap your root component with <Stage /> from "react-konva-portal" in order for Portals to work';

export type RenderFunction = () => void;

export function useForceRender(): RenderFunction {
  const [, forceRender] = useReducer(i => i + 1, 0);
  return forceRender;
}

export function zIndexComparator(a: PortalMeta, b: PortalMeta) {
  const { zIndex: zIndexA = Z_INDEX } = a;
  const { zIndex: zIndexB = Z_INDEX } = b;
  if (zIndexA === zIndexB) return 0;
  return zIndexA > zIndexB ? 1 : -1;
}

export function mountCmd(key: number, zIndex: number, children: ReactNode): MountCommand {
  return { type: 'mount', key, zIndex, children };
}

export function updateCmd(key: number, zIndex: number, children: ReactNode): UpdateCommand {
  return { type: 'update', key, zIndex, children };
}

export function unmountCmd(key: number): UnmountCommand {
  return { type: 'unmount', key };
}

export function notUnmountByKey(key: number) {
  return function (item: ManagerCommand) {
    return item.key === key && item.type !== 'unmount';
  };
}

export function warnIfDev(...args: any[]) {
  if (process.env.NODE_ENV === 'development') {
    console.warn(...args);
  }
}
