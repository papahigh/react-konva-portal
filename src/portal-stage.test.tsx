import { mount } from 'enzyme';
import Konva from 'konva';
import noop from 'lodash/noop';
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import {
  Layer as ReactKonvaLayer,
  Rect,
  Stage as ReactKonvaStage,
  StageProps as ReactKonvaStageProps,
} from 'react-konva';

import Layer from './portal-layer';
import Stage from './portal-stage';
import { PORTAL_LAYER_ID } from './utils';

describe('portal-stage.tsx', () => {
  class App extends Component<ReactKonvaStageProps> {
    public stageRef: ReactKonvaStage | null = null;
    public render() {
      return <Stage {...this.props} ref={ref => (this.stageRef = ref)} />;
    }
  }

  it('should forward ref', () => {
    const wrapper = mount<App>(<App />);
    const stageRef = wrapper.instance().stageRef;

    expect(typeof stageRef?.getStage).toEqual('function');
    expect(stageRef?.getStage()).toBeInstanceOf(Konva.Stage);
  });

  it('should pass-down props', () => {
    const wrapper = mount<App>(<App id="stage-1" x={-100} width={100} height={200} listening={false} />);
    const stage = wrapper.instance().stageRef?.getStage();

    expect(stage?.id()).toEqual('stage-1');
    expect(stage?.x()).toEqual(-100);
    expect(stage?.y()).toEqual(0);
    expect(stage?.width()).toEqual(100);
    expect(stage?.height()).toEqual(200);
    expect(stage?.listening()).toEqual(false);
  });

  it('should render portal layer', () => {
    const wrapper = mount<App>(<App />);
    const stage = wrapper.instance().stageRef?.getStage();
    const layers = stage?.getLayers() || [];

    expect(layers?.length).toEqual(1);
    expect(layers[0]?.id()).toEqual(PORTAL_LAYER_ID);
  });

  it('should handle layer components', () => {
    const wrapper = mount<App>(
      <App>
        <Layer id="host-layer" />
        <ReactKonvaLayer id="regular-layer" />
      </App>,
    );

    const stage = wrapper.instance().stageRef?.getStage();
    const layers = stage?.getLayers() || [];

    expect(layers.length).toEqual(3);
    expect(layers[0].id()).toEqual('host-layer');
    expect(layers[1].id()).toEqual('regular-layer');
    expect(layers[2].id()).toEqual(PORTAL_LAYER_ID);
  });

  it('throws error if children is not a Layer', () => {
    const _error = console.error;
    console.error = noop;
    expect(() =>
      mount(
        <App>
          <Rect fill="red" width={100} height={100} />
        </App>,
      ),
    ).toThrow('Konva error: You may only add layers to the stage.');
    console.error = _error;
  });

  it('should match snapshot', () => {
    const domElement = document.createElement('div');
    ReactDOM.render(<Stage />, domElement);
    expect(domElement).toMatchSnapshot();
  });
});
