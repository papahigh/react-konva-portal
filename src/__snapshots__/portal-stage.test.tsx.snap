// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portal-stage.tsx should match snapshot 1`] = `
<div>
  <div>
    <div
      class="konvajs-content"
      role="presentation"
      style="position: relative; user-select: none; width: 0px; height: 0px;"
    >
      <canvas
        height="0"
        style="padding: 0px; margin: 0px; border: 0px; background: transparent; position: absolute; top: 0px; left: 0px; width: 0px; height: 0px; display: block;"
        width="0"
      />
    </div>
  </div>
</div>
`;
